src/components/TypingText/TypingText.tsx

```diff
@@ -1,52 +1,238 @@
+    let currentIndex = 0;
+
+    const completionKey = STORAGE_KEYS.COMPLETION(id);
+    const progressKey = STORAGE_KEYS.PROGRESS(id);
+
+    /**
+     * Runs at the beginning of typing animation
+     * sets isTyping to false so that dot animation
+     * gets cancelled
+     */
+    const handleTypingStart = async () => {
+      if (onTypingStart) {
+        await onTypingStart(chatIndex, false);
       }
     };
 
-    typeCharacter(); // Start typing
+    /**
+     * Calculates typing delay based on progress for
+     * dynamic typing animation
+     */
+    const calculateDelay = (progress: number): number => {
+      const progressRatio = progress / normalizedText.length;
+      return speed + progressRatio * progressRatio * (TYPING_SPEEDS.MAX_SPEED - TYPING_SPEEDS.MIN_SPEED);
+    };
 
+    /**
+     * Handles typing animation completion
+     */
+    const handleTypingComplete = () => {
+      localStorage.setItem(completionKey, 'true');
+      localStorage.removeItem(progressKey);
+      onTypingComplete?.();
+    };
+
+    /**
+     * Restores typing progress from localStorage
+     */
+    const restoreProgress = () => {
+      const storedProgress = localStorage.getItem(progressKey);
+      if (storedProgress) {
+        const progress: TypingProgress = JSON.parse(storedProgress);
+        const elapsedTime = Date.now() - progress.timestamp;
+        const avgCharTime = (TYPING_SPEEDS.MIN_SPEED + TYPING_SPEEDS.MAX_SPEED) / 2;
+        const estimatedProgress = Math.floor(elapsedTime / avgCharTime);
+
+        currentIndex = Math.min(progress.index + estimatedProgress, normalizedText.length);
+      }
+      return currentIndex;
+    };
+
+    /**
+     * Main typing animation function.
+     * Saves progress and estimates further progress
+     * for simulating continues typing animation in the background
+     *  when chat is re-opened
+     */
+    const animateText = () => {
+      if (currentIndex < normalizedText.length) {
+        const delay = calculateDelay(currentIndex);
+        setDisplayedText(normalizedText.slice(0, currentIndex + 1));
+
+        // Save progress
+        const progress: TypingProgress = {
+          index: currentIndex + 1,
+          timestamp: Date.now(),
+        };
+        localStorage.setItem(progressKey, JSON.stringify(progress));
+
+        currentIndex += 1;
+        setTimeout(animateText, delay);
+      } else {
+        handleTypingComplete();
+      }
+    };
+
+    handleTypingStart();
+
+    // Check if animation is already completed
+    if (localStorage.getItem(completionKey) === 'true') {
+      setDisplayedText(normalizedText);
+    } else {
+      // Restore progress and continue animation
+      currentIndex = restoreProgress();
+      setDisplayedText(normalizedText.slice(0, currentIndex));
+
+      if (currentIndex < normalizedText.length) {
+        animateText();
+      } else {
+        handleTypingComplete();
+      }
```

Comment:
avoid having such long functions. This should be split into simple, easy to understand functions

```typescript
function orderPizza(): void {
unlockPhone()
chooseResaurant()
chooseMeal()
orderMeal()
payMeal()
closePhone()
}
```

And not 

```typescript
function orderPizza(): void {
phone.password = 'PizzePizza' 
phone.faceUnlock = true
phone.unlocking = true
phone.unlocked = false
try {
phone.forceUnlock
}
etcetctc.....
}
```